util/circular_buffer.c

```diff
@@ -141,105 +147,183 @@ int32_t cb_remove(struct circular_buffer *desc)
+int32_t cb_size(struct circular_buffer *desc, uint32_t *size)
 {
-	if (!desc)
-		return FAILURE;
+	uint32_t nb_spins;
 
-	*nb_elements =  desc->nb_elem;
+	if (!desc || !size)
+		return -EINVAL;
+
+	if (desc->write.spin_count > desc->read.spin_count)
+		nb_spins = desc->write.spin_count - desc->read.spin_count;
+	else
+		/* Integer overflow on desc->write.spin_count */
+		nb_spins = UINT32_MAX - desc->read.spin_count +
+			   desc->write.spin_count + 1;
+
+	if (nb_spins > 0)
+		*size = desc->size + desc->write.idx - desc->read.idx;
+	else
+		*size = desc->write.idx - desc->read.idx;
+
+	if (*size > desc->size) {
+		*size = desc->size;
+		return -EOVERRUN;
+	}
 
 	return SUCCESS;
 }
 
-/**
- * @brief Write data to the buffer
- * @param desc - Circular buffer reference
- * @param data - Buffer from where data is copied to the circular buffer
- * @param nb_elements - Number of elements to be copied
- * @return
- *  - \ref SUCCESS : On success
- *  - \ref FAILURE : If overflow, if nb_elements == 0 or if desc is NULL
- */
-int32_t cb_write(struct circular_buffer *desc, const void *data,
-		 uint32_t nb_elements)
+static int32_t cb_prepare_async_operation(struct circular_buffer *desc,
+		uint32_t requested_size,
+		void **buff,
+		uint32_t *raw_size_available,
+		bool is_read)
 {
-	uint32_t	to_write;
-	uint32_t	written;
-	uint32_t	writting;
-	int32_t		ret;
+	struct cb_ptr	*ptr;
+	uint32_t	available_size;
+	uint32_t	ret;
 
-	if (!desc || !nb_elements)
-		return FAILURE;
+	if (!desc || !buff || !raw_size_available)
+		return -EINVAL;
 
 	ret = SUCCESS;
+	/* Select if read or write index will be updated */
+	ptr = is_read ? &desc->read : &desc->write;
 
-	to_write = nb_elements * desc->elem_size;
-	/* Check for overflow */
-	if (to_write + (desc->nb_elem * desc->elem_size) > desc->size) {
-		/* Set the buffer as full */
-		desc->nb_elem = desc->size / desc->elem_size;
-		ret = FAILURE;
-	} else {
-		desc->nb_elem += nb_elements;
-	}
+	/* Only one transaction type possible at a single time */
+	if (ptr->async_started)
+		return -EBUSY;
 
-	written = 0;
-	while (to_write) {
-		/* Write to maximum the end of the buffer */
-		writting = min(to_write,
-			       /* Size until the end of the buffer */
-			       desc->size - (desc->write_ptr - desc->buff));
-		memcpy(desc->write_ptr, (uint8_t *)data + written, writting);
-		written += writting;
-		desc->write_ptr += writting;
-		if (desc->write_ptr - desc->buff == desc->size)
-			desc->write_ptr = desc->buff;
-		to_write -= writting;
+	if (is_read) {
+		ret = cb_size(desc, &available_size);
+		if (ret == -EOVERRUN) {
+			/* Update read index */
+			desc->read.spin_count = desc->write.spin_count - 1;
+			desc->read.idx = desc->write.idx;
+		}
+
+		/* We can only read available data */
+		requested_size = min(requested_size, available_size);
+		if (!requested_size)
+			return -EAGAIN;
 	}
 
+	/* Size to end of buffer */
+	ptr->async_size = min(requested_size, desc->size - ptr->idx);
```

Comment:
Should this take into consideration the spin_count?

Reply:
No it shouldn't. If is a read operation requested_size is set on line 213 to read a maximum of available size. If you want to write, you can write any amount of memory you want, having as limit the end of the allocated buffer (desc->size - current_idx_location)